File: app/src/main/java/com/metahumanz/pacilread/reader/modern/ReaderRuntime.java
package com.metahumanz.pacilread.reader.modern;

import android.content.Context;
import android.os.Handler;
import android.os.Looper;
import android.util.Log;

import com.metahumanz.pacilread.storage.JsonDatabase;
import com.metahumanz.pacilread.storage.SettingsStore;
import com.metahumanz.pacilread.sync.ReadingStatsSyncManager;
import com.metahumanz.pacilread.sync.WebDavClient;
import com.metahumanz.pacilread.sync.WebDavProgressSyncCoordinator;
import com.metahumanz.pacilread.tts.MimoTtsClient;
import com.metahumanz.pacilread.tts.SystemTtsClient;

import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;
import java.util.concurrent.RejectedExecutionException;

public final class ReaderRuntime {
    private static final String TAG = "PacilReadReader";

    public final ExecutorService executor = Executors.newSingleThreadExecutor();
    public final ExecutorService paginationExecutor = Executors.newSingleThreadExecutor();
    public final ExecutorService ttsExecutor = Executors.newSingleThreadExecutor();
    public final ExecutorService synthesisExecutor = Executors.newSingleThreadExecutor();
    public final Handler mainHandler = new Handler(Looper.getMainLooper());
    public final JsonDatabase databaseHelper;
    public final SettingsStore settingsStore;
    public final WebDavClient webDavClient;
    public final WebDavProgressSyncCoordinator progressSyncCoordinator;
    public final ReadingStatsSyncManager readingStatsSyncManager;
    public final MimoTtsClient mimoTtsClient;
    public final SystemTtsClient systemTtsClient;
    private volatile boolean shutdown;

    public ReaderRuntime(Context context) {
        databaseHelper = JsonDatabase.getInstance(context);
        settingsStore = new SettingsStore(context);
        webDavClient = new WebDavClient(settingsStore);
        progressSyncCoordinator = new WebDavProgressSyncCoordinator(databaseHelper, settingsStore, webDavClient);
        readingStatsSyncManager = new ReadingStatsSyncManager(context, databaseHelper, settingsStore, webDavClient);
        mimoTtsClient = new MimoTtsClient();
        systemTtsClient = new SystemTtsClient(context, settingsStore.getTtsSystemEnginePackage());
    }

    public boolean isShutdown() {
        return shutdown;
    }

    public boolean safeExecute(Runnable task, String label) {
        return safeExecute(executor, task, label);
    }

    public boolean safeExecutePagination(Runnable task, String label) {
        return safeExecute(paginationExecutor, task, label);
    }

    public boolean safeExecuteTts(Runnable task, String label) {
        return safeExecute(ttsExecutor, task, label);
    }

    public boolean safeExecuteSynthesis(Runnable task, String label) {
        return safeExecute(synthesisExecutor, task, label);
    }

    private boolean safeExecute(ExecutorService targetExecutor, Runnable task, String label) {
        if (task == null || targetExecutor == null || shutdown || targetExecutor.isShutdown()) {
            return false;
        }
        try {
            targetExecutor.execute(() -> {
                try {
                    if (!shutdown) {
                        task.run();
                    }
                } catch (RuntimeException error) {
                    Log.w(TAG, "Reader background task failed: " + safeLabel(label), error);
                }
            });
            return true;
        } catch (RejectedExecutionException error) {
            Log.d(TAG, "Reader background task rejected after shutdown: " + safeLabel(label), error);
            return false;
        }
    }

    private String safeLabel(String label) {
        return label == null || label.isBlank() ? "unnamed" : label;
    }

    public void shutdown() {
        shutdown = true;
        mainHandler.removeCallbacksAndMessages(null);
        try {
            systemTtsClient.shutdown();
        } catch (RuntimeException error) {
            Log.w(TAG, "Failed to shutdown system TTS", error);
        }
        executor.shutdownNow();
        paginationExecutor.shutdownNow();
        ttsExecutor.shutdownNow();
        synthesisExecutor.shutdownNow();
        try {
            mimoTtsClient.cancel();
        } catch (RuntimeException error) {
            Log.w(TAG, "Failed to cancel MiMo TTS", error);
        }
    }
}

File: app/src/main/kotlin/com/metahumanz/pacilread/reader/modern/ReaderRuntime.kt
package com.metahumanz.pacilread.reader.modern

import android.content.Context
import android.os.Handler
import android.os.Looper
import android.util.Log
import com.metahumanz.pacilread.storage.JsonDatabase
import com.metahumanz.pacilread.storage.SettingsStore
import com.metahumanz.pacilread.sync.ReadingStatsSyncManager
import com.metahumanz.pacilread.sync.WebDavClient
import com.metahumanz.pacilread.sync.WebDavProgressSyncCoordinator
import com.metahumanz.pacilread.tts.MimoTtsClient
import com.metahumanz.pacilread.tts.SystemTtsClient
import java.util.concurrent.Executors
import java.util.concurrent.RejectedExecutionException

class ReaderRuntime(context: Context) {
    @JvmField val executor = Executors.newSingleThreadExecutor()
    @JvmField val paginationExecutor = Executors.newSingleThreadExecutor()
    @JvmField val ttsExecutor = Executors.newSingleThreadExecutor()
    @JvmField val synthesisExecutor = Executors.newSingleThreadExecutor()
    @JvmField val mainHandler = Handler(Looper.getMainLooper())
    @JvmField val databaseHelper: JsonDatabase = JsonDatabase.getInstance(context)
    @JvmField val settingsStore = SettingsStore(context)
    @JvmField val webDavClient = WebDavClient(settingsStore)
    @JvmField val progressSyncCoordinator = WebDavProgressSyncCoordinator(databaseHelper, settingsStore, webDavClient)
    @JvmField val readingStatsSyncManager = ReadingStatsSyncManager(context, databaseHelper, settingsStore, webDavClient)
    @JvmField val mimoTtsClient = MimoTtsClient()
    @JvmField val systemTtsClient = SystemTtsClient(context, settingsStore.ttsSystemEnginePackage)
    @Volatile private var shutdown = false

    fun isShutdown(): Boolean = shutdown

    fun safeExecute(task: Runnable?, label: String?): Boolean = safeExecute(executor, task, label)
    fun safeExecutePagination(task: Runnable?, label: String?): Boolean = safeExecute(paginationExecutor, task, label)
    fun safeExecuteTts(task: Runnable?, label: String?): Boolean = safeExecute(ttsExecutor, task, label)
    fun safeExecuteSynthesis(task: Runnable?, label: String?): Boolean = safeExecute(synthesisExecutor, task, label)

    private fun safeExecute(targetExecutor: java.util.concurrent.ExecutorService?, task: Runnable?, label: String?): Boolean {
        if (task == null || targetExecutor == null || shutdown || targetExecutor.isShutdown) return false
        return try {
            targetExecutor.execute {
                try {
                    if (!shutdown) task.run()
                } catch (error: RuntimeException) {
                    Log.w(TAG, "Reader background task failed: ${safeLabel(label)}", error)
                }
            }
            true
        } catch (error: RejectedExecutionException) {
            Log.d(TAG, "Reader background task rejected after shutdown: ${safeLabel(label)}", error)
            false
        }
    }

    private fun safeLabel(label: String?): String = if (label.isNullOrBlank()) "unnamed" else label

    fun shutdown() {
        shutdown = true
        mainHandler.removeCallbacksAndMessages(null)
        try {
            systemTtsClient.shutdown()
        } catch (error: RuntimeException) {
            Log.w(TAG, "Failed to shutdown system TTS", error)
        }
        executor.shutdownNow()
        paginationExecutor.shutdownNow()
        ttsExecutor.shutdownNow()
        synthesisExecutor.shutdownNow()
        try {
            mimoTtsClient.cancel()
        } catch (error: RuntimeException) {
            Log.w(TAG, "Failed to cancel MiMo TTS", error)
        }
    }

    companion object {
        private const val TAG = "PacilReadReader"
    }
}
